cli/types/index.d.ts

```diff
@@ -4374,6 +4381,53 @@ declare namespace Cypress {
   type Diff<T extends string, U extends string> = ({ [P in T]: P } & { [P in U]: never } & { [x: string]: never })[T]
   type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>
 
+  /**
+   * Obtain the keys of T whose values are functions
+   */
+  type FunctionKeys<T> = ({ [K in keyof T]: T[K] extends (...args: any[]) => any ? K : never }[keyof T])
+
+  type ExcludeUnknownArray<T> = unknown[] extends T ? never : T
+  type ExcludeUnknown<T> = unknown extends T ? never : T
+
+  /**
+   * Obtain the union of all overloaded parameters tuples of a function type
+   *
+   * Handles functions with up to 5 overloads
+   */
+  type OverloadedParameters<T> =
+    T extends { (...args: infer A1): any; (...args: infer A2): any; (...args: infer A3): any; (...args: infer A4): any; (...args: infer A5): any; } ?
+      ExcludeUnknownArray<A1> | ExcludeUnknownArray<A2> | ExcludeUnknownArray<A3> | ExcludeUnknownArray<A4> | ExcludeUnknownArray<A5> :
+    T extends { (...args: infer A1): any; (...args: infer A2): any; (...args: infer A3): any; (...args: infer A4): any; } ?
+      ExcludeUnknownArray<A1> | ExcludeUnknownArray<A2> | ExcludeUnknownArray<A3> | ExcludeUnknownArray<A4> :
+    T extends { (...args: infer A1): any; (...args: infer A2): any; (...args: infer A3): any; } ?
+      ExcludeUnknownArray<A1> | ExcludeUnknownArray<A2> | ExcludeUnknownArray<A3> :
+    T extends { (...args: infer A1): any; (...args: infer A2): any; } ?
+      ExcludeUnknownArray<A1> | ExcludeUnknownArray<A2> :
+    T extends { (...args: infer A1): any; } ?
+      ExcludeUnknownArray<A1> :
+    any[]
+
+  type ValidReturn<TArgs, TInferredArgs, TInferredReturn> =
+    [TArgs, TInferredReturn] extends [ExcludeUnknownArray<TInferredArgs>, ExcludeUnknown<TInferredReturn>] ? TInferredReturn : never
+
+  /**
+   * Obtain the return type of a function type corresponding to the provided arguments tuple type
+   *
+   * Handles functions with up to 5 overloads
+   */
+  type OverloadedReturnType<T, TArgs extends unknown[]> =
+    T extends { (...args: infer A1): infer R1; (...args: infer A2): infer R2; (...args: infer A3): infer R3; (...args: infer A4): infer R4; (...args: infer A5): infer R5; } ?

```

Comment:
I think we can expose all this types to separate types like `AnyFunction5Args`. I think it will not be a giant problem?

Reply:
We could split into many separate types like:
```
type Process5ArgFunction<T, TArgs> =
  T extends { (...args: infer A1): infer R1; (...args: infer A2): infer R2; (...args: infer A3): infer R3; (...args: infer A4): infer R4; (...args: infer A5): infer R5; }
    ? ValidReturn<TArgs, A1, R1> | ValidReturn<TArgs, A2, R2> | ValidReturn<TArgs, A3, R3> | ValidReturn<TArgs, A4, R4> | ValidReturn<TArgs, A5, R5> : never
...
```

but I'm not sure how much that would accomplish. Is that what you mean? I think the main complication with commonization is that we need the individual inferred types.